contract/vm_state.go

```diff
@@ -0,0 +1,187 @@
+package contract
+
+import (
+	"fmt"
+	"math/big"
+	"os"
+
+	"github.com/aergoio/aergo/v2/internal/enc/base58"
+	"github.com/aergoio/aergo/v2/state"
+	"github.com/aergoio/aergo/v2/state/statedb"
+	"github.com/aergoio/aergo/v2/types"
+)
+
+type callState struct {
+	isCallback bool
+	isDeploy   bool
+
+	ctrState *statedb.ContractState
+	accState *state.AccountState
+	tx       sqlTx
+}
+
+func getCallState(ctx *vmContext, id []byte) (*callState, error) {
+	aid := types.ToAccountID(id)
+	cs := ctx.callState[aid]
+	if cs == nil {
+		bs := ctx.bs
+		accState, err := state.GetAccountState(id, bs.StateDB)
+		if err != nil {
+			return nil, err
+		}
+		cs = &callState{isCallback: true, accState: accState}
+		ctx.callState[aid] = cs
+	}
+	return cs, nil
+}
+
+func getCtrState(ctx *vmContext, id []byte) (*callState, error) {
+	cs, err := getCallState(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if cs.ctrState == nil {
+		cs.ctrState, err = statedb.OpenContractState(id, cs.accState.State(), ctx.bs.StateDB)
+	}
+	return cs, err
+}
+
+type contractInfo struct {
+	callState  *callState
+	sender     []byte
+	contractId []byte
+	rp         uint64
+	amount     *big.Int
+}
+
+func newContractInfo(cs *callState, sender, contractId []byte, rp uint64, amount *big.Int) *contractInfo {
+	return &contractInfo{
+		cs,
+		sender,
+		contractId,
+		rp,
+		amount,
+	}
+}
+
+func getOnlyContractState(ctx *vmContext, id []byte) (*statedb.ContractState, error) {
```

Comment:
This function could be moved up, just below the `getCtrState`/`getContractState` function, because they are similar